File: 005_LongestPalindromicSubstring.py
# Given a string s, return the longest 
# palindromic
 
# substring
#  in s.

 

# Example 1:

# Input: s = "babad"
# Output: "bab"
# Explanation: "aba" is also a valid answer.
# Example 2:

# Input: s = "cbbd"
# Output: "bb"


class Solution:
    def longestPalindrome(self, s: str) -> str:
        n = len(s)

        # create a dp table
        dp = [[False] * n for _ in range(n)]

        # all substring with length 1 is palindrome
        for i in range(n):
            dp[i][i] = True

        start = 0
        max_len = 1

        # check substring of length 2
        for i in range(n-1):
            if s[i] == s[i+1]:
                dp[i][i+1] = True
                start = i
                max_len = 2

        # check substring of length 3 and more
        for length in range(3,n+1):
            for i in range(n - length + 1):
                j = i + length - 1
                if dp[i+1][j-1] and s[i] == s[j]:
                    dp[i][j] = True
                    start = i
                    max_len = length

        return s[start:start+max_len]

File: 005_LongestPalindromicSubstring.cpp
// Given a string s, return the longest 
// palindromic
 
// substring
//  in s.

 

// Example 1:

// Input: s = "babad"
// Output: "bab"
// Explanation: "aba" is also a valid answer.
// Example 2:

// Input: s = "cbbd"
// Output: "bb"

class Solution {
public:
    string longestPalindrome(string s) {
        int n = s.length();

        // create a dp table
        std::vector<std::vector<bool>> dp(n, std::vector<bool>(n, false));

        int start = 0;
        int max_len = 0;

        // each variable with length 1
        for (int i = 0; i < n; ++i){
            dp[i][i] = true;
            max_len = 1;
        }

        // check length 2
        for (int i = 0; i < n - 1; ++i) {
            if (s[i] == s[i+1]){
                dp[i][i+1] = true;
                start = i;
                max_len = 2;
            }
        }

        // check length 3 and more
        for (int len = 3; len <= n; ++len){
            for (int i = 0; i <= n - len; ++ i){
                int j = i+len-1;
                if (dp[i+1][j-1] && s[i]==s[j]){
                    dp[i][j] = true;
                    start = i;
                    max_len = len;
                }    
            }
        }

        return s.substr(start, max_len);
    }
};
